posix: implement sysconf()

awojasinski opened the pull request:
The patch introduces basic implementation of sysconf() function.
It's based on macro - that means that every function call is resolved
at compile time - and is not fully complient with POSIX standard
(the errno value is not handled and passing invalid name argument
results in compilation error). Treat this commit as a starting point
for proper sysconf() implementation. The one introduced in the patch
could stay as a defult implementation.

sysconf() documentation:
https://pubs.opengroup.org/onlinepubs/9699919799/functions/sysconf.html
https://man7.org/linux/man-pages/man3/sysconf.3.html

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/56670

awojasinski committed:
posix: Add basic sysconf() function implementation

The patch introduces basic implementation of sysconf() function.
It's based on macro - that means that every function call is resolved
at compile time - and is not fully complient with POSIX standard
(the errno value is not handled and passing invalid name argument
results in compilation error). Treat this commit as a starting point
for proper sysconf() implementation. The one introduced in the patch
could stay as a defult implementation.

sysconf() documentation:
https://pubs.opengroup.org/onlinepubs/9699919799/functions/sysconf.html
https://man7.org/linux/man-pages/man3/sysconf.3.html

Fixes #56670

Signed-off-by: Adam Wojasinski <awojasinski@baylibre.com>

awojasinski committed:
tests: posix: Add test case for sysconf()

Adds test cases for sysconf() basic implementation.
Test case does not cover scenario where invalid value is passed
to the name argument of the function.

Signed-off-by: Adam Wojasinski <awojasinski@baylibre.com>

jukkar commented on include/zephyr/posix/unistd.h:
```diff
@@ -55,6 +59,153 @@ pid_t getpid(void);
+#define __z_posix_sysconf_SC_CLOCK_SELECTION		  (-1L)
+#define __z_posix_sysconf_SC_CPUTIME			  (-1L)
+#define __z_posix_sysconf_SC_FSYNC			  (-1L)
+#define __z_posix_sysconf_SC_IPV6			  (-1L)
```
The `CONFIG_NET_IPV6` tells whether IPv6 is supported or not.

jukkar reviewed (commented):
Spotted couple of network related options.

cfriedt commented on include/zephyr/posix/unistd.h:
```diff
@@ -55,6 +55,153 @@ pid_t getpid(void);
+#define __z_posix_sysconf_SC_BC_SCALE_MAX		  (-1L)
+#define __z_posix_sysconf_SC_BC_STRING_MAX		  (-1L)
+#define __z_posix_sysconf_SC_CHILD_MAX			  (-1L)
+#define __z_posix_sysconf_SC_CLK_TCK			  CONFIG_SYS_CLOCK_TICKS_PER_SEC
```
IIRC, this should be 1M on any POSIX system, regardless of what the tick rate is (I may be misremembering)

awojasinski replied:
I'll double check that, I based values of variables on the https://github.com/zephyrproject-rtos/zephyr/pull/60030 and found some that doesn't  have correct value assigned. Do you think that https://docs.zephyrproject.org/latest/services/portability/posix/conformance/index.html# is a good source of information or it's better to stick to open group specification?

awojasinski replied:
After some digging I found that's clock ticks per second. Often it's set to 100.  Someone said:
> it's the number of times the timer interrupts the CPU for scheduling and other tasks

https://stackoverflow.com/questions/19919881/sysconf-sc-clk-tck-what-does-it-return#comment29641173_19919970

cfriedt replied:
I'm not sure if this one has been resolved yet. I'm ok with it going in as-is though.

awojasinski replied:
I decided to go with `100L` as it's a commonly used value for the symbol 
https://github.com/zephyrproject-rtos/zephyr/blob/1bc840c208006d1df7bf5e9bc6416d767222dcb5/include/zephyr/posix/unistd.h#L356

cfriedt commented on include/zephyr/posix/unistd.h:
```diff
@@ -55,6 +55,153 @@ pid_t getpid(void);
 int usleep(useconds_t useconds);
 
+#ifdef CONFIG_POSIX_SYSCONF
+#define SUPPORTED_POSIX1_VERSION 200809L
```
This might be better if it were `_POSIX_VERSION`, and defined outside of `#ifdef CONFIG_POSIX_SYSCONF`.

The spec says that `_POSIX_VERSION` should be defined in `unistd.h` (newlib and picolibc define it in `sys/features.h`).

I think `#define _POSIX_VERSION 200809L` in this file should be OK though - maybe closer to the top?

cfriedt commented on include/zephyr/posix/unistd.h:
```diff
@@ -55,6 +55,153 @@ pid_t getpid(void);
+#define __z_posix_sysconf_SC_PAGE_SIZE			  (-1L)
+#define __z_posix_sysconf_SC_PAGESIZE			  (-1L)
+#define __z_posix_sysconf_SC_THREAD_DESTRUCTOR_ITERATIONS (-1L)
+#define __z_posix_sysconf_SC_THREAD_KEYS_MAX		  COND_CODE_1(CONFIG_PTHREAD_IPC,	\
```
This is great - thanks for tying these configuration bits together 😁

cfriedt commented on include/zephyr/posix/unistd.h:
```diff
@@ -55,6 +55,153 @@ pid_t getpid(void);
+								      (SUPPORTED_POSIX1_VERSION), \
+								      (-1L))
+#define __z_posix_sysconf_SC_ASYNCHRONOUS_IO		  (-1L)
+#define __z_posix_sysconf_SC_CLOCK_SELECTION		  (-1L)
```
```suggestion
#define __z_posix_sysconf_SC_CLOCK_SELECTION		  COND_CODE_1(CONFIG_POSIX_CLOCK, (_POSIX_VERSION), (-1L))
```

cfriedt commented on include/zephyr/posix/unistd.h:
```diff
@@ -55,6 +55,153 @@ pid_t getpid(void);
+#define __z_posix_sysconf_SC_MEMLOCK_RANGE		  (-1L)
+#define __z_posix_sysconf_SC_MEMORY_PROTECTION		  (-1L)
+#define __z_posix_sysconf_SC_MESSAGE_PASSING		  (-1L)
+#define __z_posix_sysconf_SC_MONOTONIC_CLOCK		  (-1L)
```
Probably an oversimplification, but for now this should be ok

```suggestion
#define __z_posix_sysconf_SC_CLOCK_SELECTION		  COND_CODE_1(CONFIG_POSIX_CLOCK, (_POSIX_VERSION), (-1L))
```

cfriedt commented on include/zephyr/posix/unistd.h:
```diff
@@ -55,6 +55,153 @@ pid_t getpid(void);
+#define __z_posix_sysconf_SC_PRIORITIZED_IO		  (-1L)
+#define __z_posix_sysconf_SC_PRIORITY_SCHEDULING	  (-1L)
+#define __z_posix_sysconf_SC_RAW_SOCKETS		  (-1L)
+#define __z_posix_sysconf_SC_READER_WRITER_LOCKS	  (-1L)
```
This would probably be more accurate
```suggestion
#define __z_posix_sysconf_SC_READER_WRITER_LOCKS		  COND_CODE_1(CONFIG_PTHREAD_IPC, (_POSIX_VERSION), (-1L))
```

although realistically, it should probably be a separate Kconfig option - maybe `POSIX_RWLOCK`? Feel free to add that as a separate commit before this commit in the same PR. Otherwise, we can do it later in a separate PR.

cfriedt commented on include/zephyr/posix/unistd.h:
```diff
@@ -55,6 +55,153 @@ pid_t getpid(void);
+#define __z_posix_sysconf_SC_SHARED_MEMORY_OBJECTS	  (-1L)
+#define __z_posix_sysconf_SC_SHELL			  (-1L)
+#define __z_posix_sysconf_SC_SPAWN			  (-1L)
+#define __z_posix_sysconf_SC_SPIN_LOCKS			  (-1L)
```
This would probably be more accurate
```suggestion
#define __z_posix_sysconf_SC_SPIN_LOCKS		  COND_CODE_1(CONFIG_PTHREAD_SPINLOCK, (_POSIX_VERSION), (-1L))
```

cfriedt commented on include/zephyr/posix/unistd.h:
```diff
@@ -55,6 +55,153 @@ pid_t getpid(void);
+#define __z_posix_sysconf_SC_MEMLOCK			  (-1L)
+#define __z_posix_sysconf_SC_MEMLOCK_RANGE		  (-1L)
+#define __z_posix_sysconf_SC_MEMORY_PROTECTION		  (-1L)
+#define __z_posix_sysconf_SC_MESSAGE_PASSING		  (-1L)
```
This would probably be more accurate
```suggestion
#define __z_posix_sysconf_SC_MESSAGE_PASSING		  COND_CODE_1(CONFIG_POSIX_MQUEUE, (_POSIX_VERSION), (-1L))
```

cfriedt commented on include/zephyr/posix/unistd.h:
```diff
@@ -55,6 +55,153 @@ pid_t getpid(void);
+#define __z_posix_sysconf_SC_THREAD_SPORADIC_SERVER	  (-1L)
+#define __z_posix_sysconf_SC_THREADS			  (-1L)
+#define __z_posix_sysconf_SC_TIMEOUTS			  (-1L)
+#define __z_posix_sysconf_SC_TIMERS			  (-1L)
```
This would probably be more accurate
```suggestion
#define __z_posix_sysconf_SC_TIMERS		  COND_CODE_1(CONFIG_POSIX_CLOCK, (_POSIX_VERSION), (-1L))
```

although likely we want to have a separate Kconfig option for it (maybe `POSIX_TIMERS`?) and also, because it was previously selected by `POSIX_CLOCK` we might need to have `POSIX_CLOCK` imply `POSIX_TIMERS`.

If you would like to make that Kconfig adjustment, feel free to do it in this PR in an earlier commit, otherwise, we can do it later at some point.

cfriedt commented on include/zephyr/posix/unistd.h:
```diff
@@ -55,6 +55,153 @@ pid_t getpid(void);
+#define __z_posix_sysconf_SC_THREAD_SAFE_FUNCTIONS	  (-1L)
+#define __z_posix_sysconf_SC_THREAD_SPORADIC_SERVER	  (-1L)
+#define __z_posix_sysconf_SC_THREADS			  (-1L)
+#define __z_posix_sysconf_SC_TIMEOUTS			  (-1L)
```
This should probably be defined to `_POSIX_VERSION` - the spec says 

> Support for timeouts for some blocking services is mandatory in POSIX.1-2017.

but that is not super descriptive of which functions need to support this option. In any case, we make every effort to support timeouts in all POSIX API calls, so it should be a given.

cfriedt commented on include/zephyr/posix/unistd.h:
```diff
@@ -55,6 +55,153 @@ pid_t getpid(void);
+#define __z_posix_sysconf_SC_THREAD_ROBUST_PRIO_PROTECT	  (-1L)
+#define __z_posix_sysconf_SC_THREAD_SAFE_FUNCTIONS	  (-1L)
+#define __z_posix_sysconf_SC_THREAD_SPORADIC_SERVER	  (-1L)
+#define __z_posix_sysconf_SC_THREADS			  (-1L)
```
This would probably be more accurate
```suggestion
#define __z_posix_sysconf_SC_THREADS		  COND_CODE_1(CONFIG_PTHREAD_IPC, (_POSIX_VERSION), (-1L))
```

cfriedt commented on include/zephyr/posix/unistd.h:
```diff
@@ -55,6 +55,153 @@ pid_t getpid(void);
+								      (SUPPORTED_POSIX1_VERSION), \
+								      (-1L))
+#define __z_posix_sysconf_SC_THREAD_CPUTIME		  (-1L)
+#define __z_posix_sysconf_SC_THREAD_PRIO_INHERIT	  (-1L)
```
This would probably be more accurate
```suggestion
#define __z_posix_sysconf_SC_PRIO_INHERIT		  _POSIX_VERSION
```

All Zephyr threads typically inherit the permissions assigned by their parent thread.

cfriedt commented on include/zephyr/posix/unistd.h:
```diff
@@ -55,6 +55,153 @@ pid_t getpid(void);
+#define __z_posix_sysconf_SC_THREAD_STACK_MIN		  PTHREAD_STACK_MIN
+#define __z_posix_sysconf_SC_THREAD_THREADS_MAX		  (-1L)
+#define __z_posix_sysconf_SC_RE_DUP_MAX			  (-1L)
+#define __z_posix_sysconf_SC_RTSIG_MAX			  (-1L)
```
@ycsin can probably comment here on `_SC_RTSIG_MAX`

ycsin replied:
https://github.com/zephyrproject-rtos/zephyr/blob/0a8d03b95f8473c8e21687adc5b26e87e4ea6c5d/lib/posix/Kconfig.signal#L23-L29

cfriedt commented on include/zephyr/posix/unistd.h:
```diff
@@ -55,6 +55,153 @@ pid_t getpid(void);
+								(CONFIG_MAX_PTHREAD_KEY_COUNT), \
+								(-1L))
+#define __z_posix_sysconf_SC_THREAD_STACK_MIN		  PTHREAD_STACK_MIN
+#define __z_posix_sysconf_SC_THREAD_THREADS_MAX		  (-1L)
```
```suggestion
#define __z_posix_sysconf_SC_THREAD_THREADS_MAX		  COND_CODE_1(CONFIG_PTHREAD_IPC, (CONFIG_MAX_PTHREAD_COUNT), (-1L))
```

cfriedt commented on include/zephyr/posix/unistd.h:
```diff
@@ -55,6 +55,153 @@ pid_t getpid(void);
+#define __z_posix_sysconf_SC_THREAD_THREADS_MAX		  (-1L)
+#define __z_posix_sysconf_SC_RE_DUP_MAX			  (-1L)
+#define __z_posix_sysconf_SC_RTSIG_MAX			  (-1L)
+#define __z_posix_sysconf_SC_SEM_NSEMS_MAX		  (-1L)
```
Seeing this tempts me to make semaphores a pooled type as well.. 🤔

cfriedt commented on include/zephyr/posix/unistd.h:
```diff
@@ -55,6 +55,153 @@ pid_t getpid(void);
+#define __z_posix_sysconf_SC_TRACE_SYS_MAX		  (-1L)
+#define __z_posix_sysconf_SC_TRACE_USER_EVENT_MAX	  (-1L)
+#define __z_posix_sysconf_SC_TYPED_MEMORY_OBJECTS	  (-1L)
+#define __z_posix_sysconf_SC_VERSION			  SUPPORTED_POSIX1_VERSION
```
Should be switched to `_POSIX_VERSION` here and elsewhere in (see other comments).

```suggestion
#define __z_posix_sysconf_SC_VERSION			  _POSIX_VERSION
```

cfriedt commented:
Generally looks very good.

@jukkar, myself and probably @ycsin have a few suggestions to maybe line up closer with what is already implemented so far.

The one thing that I think we're not addressing though, is that most of the `_SC_XXX` `sysconf()` queries mirror `_POSIX_XXX` constants or sometimes just straight defines.

So perhaps it's better to do this:

```cpp
/* define non-underscore constants */
#define PTHREAD_KEYS_MAX                COND_CODE_1(CONFIG_PTHREAD_IPC, (CONFIG_MAX_PTHREAD_KEY_COUNT), (-1L))
/* ... */

/* define _POSIX_XXX constants */
#define _POSIX_CLOCK_SELECTION        COND_CODE_1(CONFIG_POSIX_CLOCKS, (_POSIX_VERSION), (-1L))
/* ... */

/* define sysconf() variants */
#define __z_posix_sysconf_SC_THREAD_KEYS_MAX        PTHREAD_KEYS_MAX
/* ... */
#define __z_posix_sysconf_SC_CLOCK_SELECTION        _POSIX_CLOCK_SELECTION
```

Otherwise, thank you @awojasinski for taking the time to read, understand, and implement this feature! It was on the backburner for a long time 🥳 🙏

ycsin reviewed (commented):
will it be neater if we introduce an internal macro to set the sysconf value if something is enabled? i.e.:

```c
#define Z_SC_VAL_IFDEF(_def, _val) COND_CODE_1(_def, (_val), (-1L))

#define __z_posix_sysconf_SC_SEM_VALUE_MAX Z_SC_VAL_IFDEF(CONFIG_PTHREAD_IPC, CONFIG_SEM_VALUE_MAX)
```

awojasinski commented on include/zephyr/posix/unistd.h:
```diff
@@ -11,16 +11,208 @@
+#define NL_SETMAX  (255)
+#define NL_TEXTMAX (_POSIX2_LINE_MAX)
+#define NZERO	   (20)
+
+/* Runtime invariant values */
+#define AIO_LISTIO_MAX		      _POSIX_AIO_LISTIO_MAX
+#define AIO_MAX			      _POSIX_AIO_MAX
+#define AIO_PRIO_DELTA_MAX	      (0)
+#define DELAYTIMER_MAX		      _POSIX_DELAYTIMER_MAX
+#define HOST_NAME_MAX		      COND_CODE_1(CONFIG_NETWORKING,	  \
+						  (NET_HOSTNAME_MAX_LEN), \
+						  (_POSIX_HOST_NAME_MAX))
+#define LOGIN_NAME_MAX		      _POSIX_LOGIN_NAME_MAX
+#define MQ_OPEN_MAX		      _POSIX_MQ_OPEN_MAX
+#define MQ_PRIO_MAX		      _POSIX_MQ_PRIO_MAX
+#define PAGE_SIZE		      (1)
+#define PAGESIZE		      PAGE_SIZE
+#define PTHREAD_DESTRUCTOR_ITERATIONS _POSIX_THREAD_DESTRUCTOR_ITERATIONS
+#define PTHREAD_KEYS_MAX	      COND_CODE_1(CONFIG_PTHREAD_IPC,		  \
+						  (CONFIG_MAX_PTHREAD_KEY_COUNT), \
+						  (_POSIX_THREAD_KEYS_MAX))
+#define PTHREAD_THREADS_MAX	      COND_CODE_1(CONFIG_PTHREAD_IPC, CONFIG_MAX_PTHREAD_COUNT)
+#define SEM_NSEMS_MAX		      _POSIX_SEM_NSEMS_MAX
+#define SEM_VALUE_MAX		      CONFIG_SEM_VALUE_MAX
+#define SIGQUEUE_MAX		      _POSIX_SIGQUEUE_MAX
+#define STREAM_MAX		      _POSIX_STREAM_MAX
+#define SYMLOOP_MAX		      _POSIX_SYMLOOP_MAX
+#define TIMER_MAX		      CONFIG_MAX_TIMER_COUNT
+#define TTY_NAME_MAX		      _POSIX_TTY_NAME_MAX
+#define TZNAME_MAX		      _POSIX_TZNAME_MAX
+
+/* Pathname variable values */
+#define FILESIZEBITS		 (32)
+#define POSIX_ALLOC_SIZE_MIN	 (256)
+#define POSIX_REC_INCR_XFER_SIZE (1024)
+#define POSIX_REC_MAX_XFER_SIZE	 (32767)
+#define POSIX_REC_MIN_XFER_SIZE	 (1)
+#define POSIX_REC_XFER_ALIGN	 (4)
+#define SYMLINK_MAX		 _POSIX_SYMLINK_MAX
+
```
Those are definitions that should be placed in `limits.h` header file. Unfortunately that would cause compilation errors as in the `direct.h` header the `limits.h` from toolchain are included.

awojasinski commented on include/zephyr/posix/unistd.h:
```diff
@@ -11,16 +11,208 @@
+#define _POSIX_FSYNC			  (-1L)
+#define _POSIX_IPV6			  Z_SC_VAL_IFDEF(CONFIG_NET_IPV6, _POSIX_VERSION)
+#define _POSIX_JOB_CONTROL		  (-1L)
+#define _POSIX_MAPPED_FILES		  _POSIX_VERSION
```
Not sure about that but the standard says that it
 > shall always be set to the value 200809L.

cfriedt replied:
We don't support it yet unfortunately, so it should be -1L

cfriedt reviewed (commented):
Some comments. This is a lot of information to process. I think it's probably OK to go in but we should do another sweep of checks to make sure we have properly represented any features / options that are already implemented. Likely this will need some refinement still, but we have time to do that before 3.7

ycsin commented on include/zephyr/posix/unistd.h:
```diff
@@ -11,16 +11,212 @@
+#define _POSIX_MQ_PRIO_MAX		    (32)
+#define _POSIX_NAME_MAX			    (14)
+#define _POSIX_NGROUPS_MAX		    (8)
+#define _POSIX_OPEN_MAX			    CONFIG_MAX_POSIX_FDS
```
```suggestion
#define _POSIX_OPEN_MAX			    CONFIG_POSIX_MAX_FDS
```

cfriedt commented:
+@rlubos and +@pdgendt for some additional eyes. Do you see any network-related feature test macros?

rlubos commented on include/zephyr/posix/unistd.h:
```diff
@@ -11,16 +11,212 @@
+#define _POSIX_NO_TRUNC			  (-1L)
+#define _POSIX_PRIORITIZED_IO		  (-1L)
+#define _POSIX_PRIORITY_SCHEDULING	  (-1L)
+#define _POSIX_RAW_SOCKETS		  Z_SC_VAL_IFDEF(CONFIG_NETWORKING, CONFIG_NET_MAX_CONTEXTS)
```
I think it'd be better to check `CONFIG_NET_SOCKETS_PACKET`, raw sockets are not enabled by default.

jukkar commented on include/zephyr/posix/unistd.h:
```diff
@@ -11,16 +11,212 @@
+#define _XOPEN_UNIX			  (-1L)
+#define _XOPEN_UUCP			  (-1L)
+
+/* Maximum values */
```
Should this say `Minimum values`

awojasinski replied:
If you look into https://pubs.opengroup.org/onlinepubs/009695399/basedefs/limits.h.html the meaning of maximum and minimum is inverted. I was using the document as a reference for the implementation

jukkar commented on include/zephyr/posix/unistd.h:
```diff
@@ -11,16 +11,212 @@
+/* Maximum values */
+#define _POSIX_CLOCKRES_MIN (20000000L)
+
+/* Minimum values */
```
and this `Maximum values`

jukkar commented on include/zephyr/posix/unistd.h:
```diff
@@ -11,16 +11,212 @@
+#define MQ_OPEN_MAX		      _POSIX_MQ_OPEN_MAX
+#define MQ_PRIO_MAX		      _POSIX_MQ_PRIO_MAX
+#ifndef PAGE_SIZE
+	#define PAGE_SIZE	      (1)
```
I think we do not normally indent defines, so this looks like it is out of place

awojasinski replied:
I thought that putting the `PAGE_SIZE` symbol two times with the same indentation in such a big block would be less readable. If you think it's better to remove indentation that's fine to me

pdgendt replied:
For readability we usually put newlines around such a block

```c
// ...
#define MQ_PRIO_MAX		_POSIX_MQ_PRIO_MAX

#ifndef PAGE_SIZE
#define PAGE_SIZE		(1)
#endif

#ifndef PAGESIZE
#define PAGESIZE		PAGE_SIZE
#endif

#define PTHREAD_DESTRUCTOR_ITERATIONS _POSIX_THREAD_DESTRUCTOR_ITERATIONS
// ...
```

ycsin commented on include/zephyr/posix/unistd.h:
```diff
@@ -11,16 +11,212 @@
+#define MQ_PRIO_MAX		      _POSIX_MQ_PRIO_MAX
+#ifndef PAGE_SIZE
+	#define PAGE_SIZE	      (1)
+#endif
```
for readability:

```suggestion
#endif /* PAGE_SIZE */
```

cfriedt replied:
@ycsin - is this a blocking suggestion?

awojasinski replied:
IMO the comment is not needed as `#ifndef` is two lines above

ycsin commented on include/zephyr/posix/unistd.h:
```diff
@@ -55,6 +251,137 @@ pid_t getpid(void);
+#define __z_posix_sysconf_SC_TTY_NAME_MAX		  TTY_NAME_MAX
+#define __z_posix_sysconf_SC_TZNAME_MAX			  TZNAME_MAX
+
+#define sysconf(x) (long)UTIL_EVAL(__z_posix_sysconf ## x)
```
`CONCAT` in `util.h` might be better suited for this

```suggestion
#define sysconf(x) (long)CONCAT(__z_posix_sysconf, x)
```

https://github.com/zephyrproject-rtos/zephyr/blob/592647cf0b8adb713bdd98bc7e1d67986fda4d3b/include/zephyr/sys/util.h#L271-L283

The `UTIL_EVAL` is in the `util_internal.h`, which provides _'helper macros needed by sys/util.h'_

ycsin commented on include/zephyr/posix/unistd.h:
```diff
@@ -11,16 +11,217 @@
+#define PTHREAD_KEYS_MAX	      COND_CODE_1(CONFIG_PTHREAD_IPC,		  \
+						  (CONFIG_MAX_PTHREAD_KEY_COUNT), \
+						  (_POSIX_THREAD_KEYS_MAX))
+#define PTHREAD_THREADS_MAX	      COND_CODE_1(CONFIG_PTHREAD_IPC, CONFIG_MAX_PTHREAD_COUNT)
```
`COND_CODE_1` expects 3 args

awojasinski committed:
doc: posix: Update POSIX supported API documentation with sysconf()

Updates in documentation support for `sysconf` API
in the `POSIX_SINGLE_PROCESS` group option.

Signed-off-by: Adam Wojasinski <awojasinski@baylibre.com>

pdgendt commented on include/zephyr/posix/unistd.h:
```diff
@@ -11,16 +11,217 @@
+#define PTHREAD_KEYS_MAX	      COND_CODE_1(CONFIG_PTHREAD_IPC,		  \
+						  (CONFIG_MAX_PTHREAD_KEY_COUNT), \
+						  (_POSIX_THREAD_KEYS_MAX))
+#define PTHREAD_THREADS_MAX	      COND_CODE_1(CONFIG_PTHREAD_IPC, CONFIG_MAX_PTHREAD_COUNT, 0)
```
`COND_CODE_1` requires that it's arguments have parenthesis

```suggestion
#define PTHREAD_THREADS_MAX	      COND_CODE_1(CONFIG_PTHREAD_IPC, (CONFIG_MAX_PTHREAD_COUNT), (0))
```

cfriedt commented on include/zephyr/posix/unistd.h:
```diff
@@ -11,16 +11,219 @@
+#define MQ_PRIO_MAX		      _POSIX_MQ_PRIO_MAX
+
+#ifndef PAGE_SIZE
+#define PAGE_SIZE		      (1)
```
Oh wow - this definitely needs a Kconfig option now 😅 

Would suggest `CONFIG_POSIX_PAGE_SIZE_BITS` and a top-level entry in `lib/posix/Kconfig`

```kconfig
config POSIX_PAGE_SIZE_BITS
	int "Number of bits to use for PAGE_SIZE"
	range 6 16
	default 12 if POSIX_API
	default 6
	help
	  Define PAGE_SIZE as BIT(n), where n is the value configured here.
	  
	  PAGE_SIZE is supported in the range [64, 65536]
	  
	  If CONFIG_POSIX_API=y, this PAGE_SIZE defaults to 4096, otherwise, it is 64 bytes.
```

With that we can have
```cpp
#define PAGE_SIZE BIT(CONFIG_POSIX_PAGE_SIZE_BITS)
```

awojasinski replied:
Unfortunately there are few platforms that define `PAGE_SIZE`, also tf-m and tf-a is using the symbol. Initially I made that mistake and didn't use preprocessor guards for the `PAGE_SIZE` symbol and that resulted in failing tests. If you want I can remove `#ifdef`/`#endif` for the symbol and we can observe what will happen

cfriedt replied:
Works for me - thanks @awojasinski

cfriedt commented:
@jukkar, @rlubos, @pdgendt - any additional change requests?

pdgendt commented:
> @jukkar, @rlubos, @pdgendt - any additional change requests?

I don't see any obvious issues, but not that familiar with the functionality either so not sure if my approval is representative.
